### PR TITLE
Add PoleReluSystem.

### PR DESCRIPTION
**Pull request by hongkai-dai**
The system approximates the dynamics of pole/end-effector system using ReLU networks

---
This change is [Reviewable](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401)

### CODE REVIEW

**Review by hongkai-dai — COMMENTED** [review]
+@lujieyang for review please, thanks!

> *[Reviewable](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-:-MqAvJ2O0LIFgsZDjecO:b-ioxum3)* status: 0 of 2 files reviewed, all discussions resolved (waiting on @lujieyang)

**Comment by lujieyang** [comment]
___
*[neural_network_lyapunov/examples/pole/pole_relu_system.py, line 17 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqCtwVyCSk5sImj5eNP:-MqCtwVyCSk5sImj5eNQ:b-ei0zpn) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/pole_relu_system.py#L17)):*
> ```Python
>     [ẋ_A[n+1] ẏ_A[n+1] ż_A[n+1] ẋ_AB[n+1] ẏ_AB[n+1]] =
>     [ẋ_A[n] ẏ_A[n] ż_A[n] ẋ_AB[n] ẏ_AB[n]] +
>     ϕ(x_AB[n], y_AB[n], ẋ_AB[n], ẏ_AB[n], u[n]) -
> ```

\phi is absorbing dt. Therefore, we have to train \phi for different dt.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401)* status: 0 of 2 files reviewed, 1 unresolved discussion (waiting on @lujieyang)

___
*[neural_network_lyapunov/examples/pole/pole_relu_system.py, line 17 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqCtwVyCSk5sImj5eNP:-MqCvNHc1Lg1tccvLthx:br506tt) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/pole_relu_system.py#L17)):*
<details><summary><i>Previously, lujieyang wrote…</i></summary><blockquote>

\phi is absorbing dt. Therefore, we have to train \phi for different dt.
</blockquote></details>

Right and that is by intention. I think you are suggesting to represent the continuous dynamics xdot = f(x, u) using a network phi. But what we want is the discrete dynamics x[n+1]. If we use first-order explicit Euler integration as x[n+1] = x[n] + xdot[n] * dt then it is true we only need to learn xdot = f(x, u), and then multiplies dt. But first order explicit Euler integration has very bad accuracy, so it is more proper to use higher order integrations. Then it is not sufficient to just learn xdot = f(x, u), but to learn the discrete dynamics x[n+1] = F(x[n], u[n]) directly, which handles the time integration inside the dynamics update F(x[n], u[n]).

**Comment by lujieyang** [comment]
___
*[neural_network_lyapunov/examples/pole/pole_relu_system.py, line 133 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqFH8iC1JYSr_U7JhKP:-MqFH8iC1JYSr_U7JhKQ:b-ybaerk) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/pole_relu_system.py#L133)):*
> ```Python
>                                self.u_equilibrium))),
>             sense=gurobipy.GRB.EQUAL,
>             name="quadrotor_dynamics_output")
> ```

pole_dynamics_output

**Comment by lujieyang** [comment]
___
*[neural_network_lyapunov/examples/pole/pole_relu_system.py, line 17 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqCtwVyCSk5sImj5eNP:-MqFKLDjD4aSpiujLcuT:bz2xonu) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/pole_relu_system.py#L17)):*
<details><summary><i>Previously, hongkai-dai (Hongkai Dai) wrote…</i></summary><blockquote>

Right and that is by intention. I think you are suggesting to represent the continuous dynamics xdot = f(x, u) using a network phi. But what we want is the discrete dynamics x[n+1]. If we use first-order explicit Euler integration as x[n+1] = x[n] + xdot[n] * dt then it is true we only need to learn xdot = f(x, u), and then multiplies dt. But first order explicit Euler integration has very bad accuracy, so it is more proper to use higher order integrations. Then it is not sufficient to just learn xdot = f(x, u), but to learn the discrete dynamics x[n+1] = F(x[n], u[n]) directly, which handles the time integration inside the dynamics update F(x[n], u[n]).
</blockquote></details>

Is there advantage of learning residual dynamics over the entire forward dynamics?

**Comment by lujieyang** [comment]
___
*[neural_network_lyapunov/examples/pole/test/test_pole_relu_system.py, line 54 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqFLYZxAx2FJbkpB0U7:-MqFLYZxAx2FJbkpB0U8:b-65rcpz) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/test/test_pole_relu_system.py#L54)):*
> ```Python
>             x_next_i = dut.step_forward(x_samples[i], u_samples[i])
>             np.testing.assert_allclose(x_next[i].detach().numpy(),
>                                        x_next_i.detach().numpy())
> ```

I'm not quite seeing the purpose of checking that they are the same. It seems they are going through the same function with one in batch and another with single data point.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @hongkai-dai and @lujieyang)

___
*[neural_network_lyapunov/examples/pole/pole_relu_system.py, line 17 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqCtwVyCSk5sImj5eNP:-MqFbsMkDKj0O44aax2L:b-crdsl2) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/pole_relu_system.py#L17)):*
<details><summary><i>Previously, lujieyang wrote…</i></summary><blockquote>

Is there advantage of learning residual dynamics over the entire forward dynamics?
</blockquote></details>

Yes, if we learn the entire forward dynamics, then the input to the network should have size 10, including x_AB, y_AB, xdot_A, ydot_A, zdot_A, xdot_AB, ydot_AB, zdot_AB, u. On the other hand, if we learn residue dynamics, the network input size is 7, it doesn't need to include xdot_A, ydot_A, zdot_A.
___
*[neural_network_lyapunov/examples/pole/pole_relu_system.py, line 133 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqFH8iC1JYSr_U7JhKP:-MqFcKi-38VdIqZ-HlCP:b-896fix) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/pole_relu_system.py#L133)):*
<details><summary><i>Previously, lujieyang wrote…</i></summary><blockquote>

pole_dynamics_output
</blockquote></details>

Done.

**Review by hongkai-dai — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401)* status: 0 of 2 files reviewed, 2 unresolved discussions (waiting on @lujieyang)

___
*[neural_network_lyapunov/examples/pole/test/test_pole_relu_system.py, line 54 at r1](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401#-MqFLYZxAx2FJbkpB0U7:-MqFcWl65Y3yinXlmHyp:bz7rnyr) ([raw file](https://github.com/stanfordasl/neural-network-lyapunov/blob/18d724f130595ddf6f801301af22acc84f140913/neural_network_lyapunov/examples/pole/test/test_pole_relu_system.py#L54)):*
<details><summary><i>Previously, lujieyang wrote…</i></summary><blockquote>

I'm not quite seeing the purpose of checking that they are the same. It seems they are going through the same function with one in batch and another with single data point.
</blockquote></details>

They come from two branches in the same function. There is an if branch inside `step_forward` function, depending on if the input `x, u` is a single state/control or a batch of states/controls. Hence I want to check if the two branches compute the same result.

**Review by lujieyang — APPROVED** [review]
> Reviewed 2 of 2 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/stanfordasl/neural-network-lyapunov/401)* status: :shipit: complete! all files reviewed, all discussions resolved (waiting on @hongkai-dai)
